lib/Controller/WopiController.php

```diff
@@ -0,0 +1,873 @@
+		$this->logger->debug('PutFile: WOPI header timestamp: {wopiHeaderTime}', [
+			'app' => $this->appName,
+			'wopiHeaderTime' => $wopiHeaderTime
+		]);
+
+		// get owner and editor uid's
+		$owner = $res['owner'];
+		$editor = $res['editor'];
+
+		$file = $this->fileService->getFileHandle($fileId, $owner, $editor);
+		if (!$file) {
+			$this->logger->warning('PutFile: Could not retrieve file', ['app' => $this->appName]);
+			return new JSONResponse([], Http::STATUS_NOT_FOUND);
+		}
+
+		// Handle wopiHeaderTime
+		if (!$wopiHeaderTime) {
+			$this->logger->debug('PutFile: X-LOOL-WOPI-Timestamp absent. Saving file.', ['app' => $this->appName]);
+		} elseif ($wopiHeaderTime != Helper::toISO8601($file->getMTime())) {
+			$this->logger->debug('PutFile: Document timestamp mismatch ! WOPI client says mtime {headerTime} but storage says {storageTime}', [
+				'app' => $this->appName,
+				'headerTime' => $wopiHeaderTime,
+				'storageTime' => Helper::toISO8601($file->getMtime())
+			]);
+			// Tell WOPI client about this conflict.
+			return new JSONResponse(['LOOLStatusCode' => self::LOOL_STATUS_DOC_CHANGED], Http::STATUS_CONFLICT);
+		}
+
+		// Read the contents of the file from the POST body and store.
+		$content = \fopen('php://input', 'r');
+		$this->logger->debug(
+			'PutFile: storing file {fileId}, editor: {editor}, owner: {owner}.',
+			[
+				'app' => $this->appName,
+				'fileId' => $fileId,
+				'editor' => $editor,
+				'owner' => $owner
+			]
+		);
+		$file->putContent($content);
+
+		$this->logger->debug('PutFile: mtime', ['app' => $this->appName]);
+
+		$mtime = $file->getMtime();
+
+		return new JSONResponse([
+			'status' => 'success',
+			'LastModifiedTime' => Helper::toISO8601($mtime)
+		], Http::STATUS_OK);
+	}
+
+	/**
+	 * The Files endpoint operation PutFileRelative.
+	 */
+	public function wopiPutFileRelative(string $documentId): JSONResponse {
+		$token = $this->request->getParam('access_token');
+
+		list($fileId, , , ) = Helper::parseDocumentId($documentId);
+		$this->logger->debug('PutFileRelative: file {fileId}, token {token}.', [
+			'app' => $this->appName,
+			'fileId' => $fileId,
+			'token' => $token]);
+
+		$row = new Db\Wopi();
+		$row->loadBy('token', $token);
+
+		$res = $row->getWopiForToken($token);
+		if ($res == false) {
+			$this->logger->debug('PutFileRelative: get token failed.', ['app' => $this->appName]);
+			return new JSONResponse([], Http::STATUS_FORBIDDEN);
+		}
+
+		$canWrite = $res['attributes'] & WOPI::ATTR_CAN_UPDATE;
+		if (!$canWrite) {
+			$this->logger->debug('PutFileRelative: not allowed.', ['app' => $this->appName]);
+			return new JSONResponse([], Http::STATUS_FORBIDDEN);
+		}
+
+		// get owner and editor uid's
+		$owner = $res['owner'];
+		$editor = $res['editor'];
+
+		// Retrieve suggested target
+		$suggested = $this->request->getHeader('X-WOPI-SuggestedTarget');
+		$suggested = \iconv('utf-7', 'utf-8', $suggested);
```

Comment:
I think we'll need a comment here. Unless it's strictly needed, I'd rather not touch the encoding.
Even better if you can provide a link to an official source to prove that utf-7 is used.

Reply:
this was copied over, would prefer not to refactor this part as I already refactored a lot